tiny_dnn/core/framework/tensor.h

```diff
@@ -0,0 +1,156 @@
+    The project versioning (Git) records all such contribution source information.
+
+    LICENSE
+
+    The BSD 3-Clause License
+
+
+    Redistribution and use in source and binary forms, with or without
+    modification, are permitted provided that the following conditions are met:
+
+    * Redistributions of source code must retain the above copyright notice, this
+      list of conditions and the following disclaimer.
+
+    * Redistributions in binary form must reproduce the above copyright notice,
+      this list of conditions and the following disclaimer in the documentation
+      and/or other materials provided with the distribution.
+
+    * Neither the name of tiny-dnn nor the names of its
+      contributors may be used to endorse or promote products derived from
+      this software without specific prior written permission.
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+    AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+    DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+    FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+    DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+    SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+    OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+    OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+#pragma once
+
+namespace tiny_dnn {
+
+/* Class modelling a Tensor
+ */
+class Tensor {
+ public:
+    /* Default constructor for the Tensor class
+     * @param batch The size of the batch
+     * @param width The width of the tensor
+     * @param heigth The height of the tensor
+     * @param depth The number of channels
+     */
+    explicit Tensor(const cnn_size_t batch,
+                    const cnn_size_t width,
+                    const cnn_size_t height,
+                    const cnn_size_t depth) : shape_(4) {
+        init_data (batch, width, height, depth);
+        init_shape(batch, width, height, depth);
+    }
+    
+    // Returns the tensor shape
+    std::vector<cnn_size_t> shape() const { return shape_; }
+    
+    // Returns the value of a requested position in the tensor
+    float_t at(const cnn_size_t batch,  const cnn_size_t width,
+               const cnn_size_t height, const cnn_size_t depth) const {
+        return *access_data(batch, width, height, depth);
+    }
+
+    // Returns the pointer to a requested position in the tensor
+    float_t* ptr(const cnn_size_t batch,  const cnn_size_t width,
+                 const cnn_size_t height, const cnn_size_t depth) const {
+        return access_data(batch, width, height, depth);
+    }
```

Comment:
@edgarriba 
const method should not return non-const pointer. It is problematic if we can change the internal data via const method (it will break the coherency between cpu/gpu data). I want to have const/non-const overloading, and zero-overhead access like ```std::vector```.

```cpp
template <class T>
class vector {
public:
    // checked version (throw exceptions for out-of-range error)
    T& at(size_type index);
    const T& at(size_type index) const;

    // zero-overhead version (same performance to raw pointer access. have an assertion for out-of-range error)
    T& operator [] (size_type index);
    const T& operator [] (size_type index) const;
}; 
```